server/utils/ThreadObserver.py

```diff
@@ -0,0 +1,84 @@
+import psycopg2
+from apiclient.discovery import build
+from apiclient import errors
+import httplib2
+from config.default import (
+    REDIS_URL,
+    DATABASE_URL
+)
+import redis
+from rq import (
+    Queue,
+    get_current_job
+)
+
+
+def update_thread(thread_id, credentials, is_final_job):
+    """Update the status of a single thread
+    
+    return the Redis job
+    """
+    # get number of replies for this thread
+    service = build(
+        'gmail', 'v1', 
+        http=credentials.authorize(http = httplib2.Http())
+    )
+    try:
+        thread = service.users().threads().get(userId="me", id=thread_id).execute()
+        messages = thread['messages']
+        thread_id = thread['id']
+        print(f'thread id: {thread_id} - number of messages in this thread: {len(messages)}')
+        if is_replied_to(thread):
+            set_replied_to(thread)
+        if is_final_job:
+            complete_thread_job(get_current_job())
+    except (errors.HttpError, errors.Error) as e:
+        print('ThreadHelper: An error occurred: %s' % e)
+
+def enqueue_thread_update(thread_id, credentials, is_final_job):
+    conn = redis.from_url(REDIS_URL)
+    q = Queue(connection=conn)
+    job = q.enqueue(
+        update_thread,
+        thread_id,
+        credentials,
+        is_final_job
+    )
+    return job
+
+def set_replied_to(thread):
+    """Mark an entry in the threads table complete in the database"""
+    conn = psycopg2.connect(DATABASE_URL)
+    cur = conn.cursor()
+    cur.execute("""
+    UPDATE threads
+    SET replied_to=true
+    WHERE id=%s
+    """,
+    (thread['id'],))
+    
+    conn.commit()
+    cur.close()
+    conn.close()
+
+def complete_thread_job(job):
+    """Mark a thread_task entry complete in the database"""
+    conn = psycopg2.connect(DATABASE_URL)
+    cur = conn.cursor()
+    cur.execute("""
+    UPDATE thread_tasks
+    SET complete=true
+    WHERE id=%s
+    """,
+    (job.id,))
+
+    conn.commit()
+    cur.close()
+    conn.close()
+
+def is_replied_to(thread):
+    """Return true if a thread has more than 1 message in it"""
+    if len(thread['messages']) > 1:
+        return True
+    return False
+
```

Comment:
What if the sender sends a followup message, or double send?

Reply:
It would be in the same thread, but I couldn’t find a way to determine who sent each message in a thread without making more requests to the gmail api, that information just isn’t there in the response. I thought one request for every thread for every user that logs in was already a lot, so I left it at that.